feat: Add `shinylive url` command

gadenbuie opened the pull request:
Adds `shinylive url`, a command group to `encode` (or create) a shinylive.io URL for Python or R apps from local files or `decode` the existing URL. The core functionality is wrapped in `encode_shinylive_url()` and `decode_shinylive_url()` which are exported to facilitate creating and decoding shinylive URLs from within a Python session (the primary use case is for the Component and Layout galleries, but I'm sure this would be broadly useful).

```
❯ shinylive url encode --help
Usage: shinylive url encode [OPTIONS] APP [FILES]...

  Create a shinylive.io URL for a Shiny app from local files.

  APP is the path to the primary Shiny app file.

  FILES are additional supporting files or directories for the app.

  On macOS, you can copy the URL to the clipboard with:

      shinylive url encode app.py | pbcopy

Options:
  -m, --mode [editor|app]         The shinylive mode: include the editor or
                                  show only the app.  [required]
  -l, --language [python|py|R|r]  The primary language used to run the app, by
                                  default inferred from the app file.
  -v, --view                      Open the link in a browser.
  --no-header                     Hide the Shinylive header.
  --help                          Show this message and exit.
```

```
❯ shinylive url decode --help
Usage: shinylive url decode [OPTIONS] [URL]

  Decode a shinylive.io URL.

  URL is the shinylive editor or app URL. If not specified, the URL will be
  read from stdin, allowing you to read the URL from a file or the clipboard.

  On macOS, you can read the URL from the clipboard with:

      pbpaste | shinylive url decode

Options:
  --out TEXT  Output directory into which the app's files will be written. The
              directory is created if it does not exist.
  --json      Prints the decoded shinylive bundle as JSON to stdout, ignoring
              --out.
  --help      Show this message and exit.
```

## Examples

Copy the [multiple files Python app example link](https://shinylive.io/py/editor/#code=NobwRAdghgtgpmAXGKAHVA6VBPMAaMAYwHsIAXOcpMAMwCdiYACAZwAsBLCbJjmVYnTJMAgujxM6lACZw6EgK4cAOhHqMmCshwA2LXv0HCWARwVQpq1WlQB9JUwC8mjligBzOLZo6l0gBSqTMEuGFyoWrYsOhyydIGQyvhMSQBySRIADBIAjJnZTABMmQCUeEEhShjEWhFkthQAHvUAbnIARlDaMAlkzUllqiVWEKqyNKxybfHhWhI1ZHUSLHAsLBykJYgVwQACC3U7TLtSEHEYTWRH40x9ZP5bRyFMLVA6TqxmFnD+s2QYEAewwgz2eUjICjoIJoSRAfwBDwAvp9zFJpLx9CBXjpEUkRtZ0B8xKh-DZ7BxllM5BJZO0FO5HAAVOgKODAsCIvDgaDwahaXQsLC4AgkciUMjUG6mVE-CCPEEhcGQkEggBUTFGEA5AF0gA):

```
pbpaste | shinylive url decode
```
```
## file: app.py

from shiny import App, render, ui
from utils import square

app_ui = ui.page_fluid(
    ui.input_slider("n", "N", 0, 100, 20),
    ui.output_text_verbatim("txt"),
)


def server(input, output, session):
    @output
    @render.text
    def txt():
        val = square(input.n())
        return f"{input.n()} squared is {val}"


app = App(app_ui, server, debug=True)

## file: utils.py

def square(n):
    return n * n
```

Copy this example Shiny app

```python
from shiny import App, render, ui

app_ui = ui.page_fluid(
    ui.input_slider("x", "Slider input", min=0, max=20, value=10),
    ui.output_text_verbatim("txt"),
)


def server(input, output, session):
    @output
    @render.text
    def txt():
        return f"x: {input.x()}"


app = App(app_ui, server, debug=True)
```

and then generate a link and open it on shinylive

```
pbpaste | shinylive url encode --view
```

Try the same idea, but with this R app

```r
library(shiny)

ui <- fluidPage(
  h2(textOutput("currentTime"))
)

server <- function(input, output, session) {
  output$currentTime <- renderText({
    invalidateLater(1000, session)
    paste("The current time is", Sys.time())
  })
}

shinyApp(ui, server)
```

```
pbpaste | shinylive url encode --view
```

Or apply the same idea to a local app on your computer. In this case, the language is inferred from the file extension.

```
shinylive url encode -v posit-dev/py-shiny/examples/event/app.py

shinylive url encode -v rstudio/shiny/inst/examples/11_timer/app.R
```

gadenbuie commented on shinylive/_url.py:
```diff
@@ -0,0 +1,96 @@
+
+
+def make_shinylive_url(
+    app: str | Path,
```
Here `app` is a path to a file on disk, but I can see utility in allowing `app` to be the actual code of a single `app.py` or `app.R` file. That doesn't need to make it into this PR but I'd be happy for advice about how to design the API in a pythonic way.

gadenbuie replied:
Here `app` is a path to a file on disk, but I can see utility in allowing `app` to be the actual code of a single `app.py` or `app.R` file. That doesn't need to make it into this PR but I'd be happy for advice about how to design the API in a pythonic way.

gadenbuie commented on shinylive/_url.py:
```diff
@@ -0,0 +1,96 @@
+        The mode of the application. Defaults to "editor".
+    language
+        The language of the application. Defaults to None.
+    no_header
+        Whether to include a header. Defaults to False.
```
This follows the CLI interface, where `--no-header` disables the header. I'd be happy to reverse this for the Python API and have `header=True` by default.

gadenbuie commented on shinylive/_url.py:
```diff
@@ -0,0 +1,96 @@
+    type: Literal["text", "binary"]
+
+
+def read_file(file: str | Path, root_dir: str | Path | None = None) -> FileContentJson:
```
```suggestion
# Copied from https://github.com/posit-dev/py-shiny/blob/main/docs/_renderer.py#L231
def read_file(file: str | Path, root_dir: str | Path | None = None) -> FileContentJson:
```

gadenbuie commented on setup.cfg:
```diff
@@ -36,6 +36,7 @@ install_requires =
     shiny
     click>=8.1.7
     appdirs>=1.4.4
+    lzstring>=1.0.4
```
I added `lzstring` here but the checks are still reporting "no module named 'lzstring'". I can't repro this locally, `pip install -e ".[dev,test]"` installs lzstring.

cpsievert commented on shinylive/_main.py:
```diff
@@ -469,6 +471,154 @@ def defunct_help(cmd: str) -> str:
+    help="""
+Create a shinylive.io URL for a Shiny app from local files.
+
+APP is the path to the primary Shiny app file.
+
+FILES are additional supporting files for the app.
```
If there are lots of supporting files, this could be tedious to specify from command line? Seems like we should take an `APPDIR` similar to `shinylive export` and `rsconnect::deploytApp()`?

gadenbuie replied:
Good idea, I've set it up now so that if a file in FILES is a directory, all of the files in that directory will be included in the bundle (recursively).

gadenbuie commented on shinylive/_main.py:
```diff
@@ -469,6 +471,153 @@ def defunct_help(cmd: str) -> str:
+""",
+)
+@click.option(
+    "--out",
```
This should be `--dir` for consistency

```suggestion
    "--dir",
```

gadenbuie commented on shinylive/_url.py:
```diff
@@ -0,0 +1,187 @@
+
+    # if app has a newline, then it's app content, not a path
+    if isinstance(app, str) and "\n" in app:
+        # now language is required
+        if language is None:
+            raise ValueError("If `app` is a string, then `language` must be specified.")
```
We could also inspect `app` for `import shiny`, `from shiny` or `library(shiny)` to decide, before bailing.

wch commented on shinylive/_url.py:
```diff
@@ -0,0 +1,217 @@
+
+
+def encode_shinylive_url(
+    app: str | Path,
```
I feel uneasy about allowing `app` to be a string which is either a path, or the content of an app.py/app.R file. This kind of dual purpose of a parameter feels weird, and it has bitten us in the past (IIRC, with jsonlite).

I think there should just be a different parameter, like `app_contents`, for that case.

gadenbuie replied:
In light of also wanting to add an "encode to JSON" option, I think it makes sense to break apart the work being done here in `encode_shinylive_url()` to break it into two steps of creating the `list[FileContentJson]` bundle and to encode the bundle directly into a URL. With that pairing, we could have separate functions that perform each task rather than overloading a single argument.

At that point, we'd be pushing this check onto users to decide which function to call, but that doesn't seem terribly cumbersome.

gadenbuie replied:
> I feel uneasy about allowing `app` to be a string which is either a path, or the content of an app.py/app.R file. This kind of dual purpose of a parameter feels weird

I agree that the dual purpose `app` parameter feels weird, but I feel equally or slightly more weird about two competing required parameters. If we added `app_contents`, then only one of `app` or `app_contents` would be allowed, since they would both serve the purpose of _the primary app_.

In the latest commit, I've broken this into a two steps:

1. First choose `create_shinylive_bundle_file()` or `create_shinylive_bundle_text()` depending on whether you want to pass `app` as the path to an app **file** or `app` as the **text** contents of an app.
  
   Both functions return an _app bundle_ which is just a dictionary containing the list of `FileContentJson` as `files`, plus the `language` since we might have needed to detect it from the value of `app`.

2. Then pass this bundle to `create_shinylive_url()` where there a few additional options, such as `mode` and `header`.

This certainly feels more composable to me, and I like the idea of being able to give `create_shinylive_url()` a fully specified app bundle (where I as the user have complete control over the contents).

But I'm not convinced this leads to a better user experience. As an end-user of these functions, I can see this approach pushing some friction into my workflow. I think there's utility in retaining an `encode_shinylive_url()` that matches the API of the CLI command and that does all of the "magic" for the user. Would this be a reasonable middle ground?

gadenbuie replied:
I relocated the app language detection so that we don't need a new app bundle type, we can just use `list[FileContentJson]` for the app payload.

I also added `encode_shinylive_url()`, which now serves as a programmatic sibling of the CLI command, while retaining the individual functions for each step.

wch commented on shinylive/_main.py:
```diff
@@ -469,6 +471,153 @@ def defunct_help(cmd: str) -> str:
+        import webbrowser
+
+        webbrowser.open(url)
+    return
```
```suggestion
```

wch commented on shinylive/_main.py:
```diff
@@ -469,6 +471,153 @@ def defunct_help(cmd: str) -> str:
+                print("## type: binary")
+            print("")
+            print(file["content"].encode("utf-8", errors="ignore").decode("utf-8"))
+
+    return
```
```suggestion
```

wch commented on shinylive/_main.py:
```diff
@@ -469,6 +471,153 @@ def defunct_help(cmd: str) -> str:
+)
+@click.argument("app", type=str, nargs=1, required=True, default="-")
+@click.argument("files", type=str, nargs=-1, required=False)
+def encode(
```
A `--json` option would be useful for encoding.

gadenbuie replied:
I'm curious how you see the JSON option being useful, but I completely agree that in making this possible we will definitely push to code towards a better structure.

wch replied:
I've had a few occasions where I've used `shinylive export` and then made tweaks to the generated `app.json` file, and it might have been easier and quicker to be able to simply re-export the JSON from this command. But that is a pretty niche debugging use case, and not particularly important to do at this point.

gadenbuie replied:
It's done :smile

gadenbuie replied:
Wait, I think it's done. Right now it outputs the app bundle as JSON, i.e. taking the app and supporting file inputs and making their shinylive JSON representation. It occurred to me just now that maybe you'd want to take JSON inputs and return a shinylive URL. If that's the case, let me know, otherwise if you wanted JSON output we're all set.

wch commented on shinylive/_main.py:
```diff
@@ -469,6 +471,153 @@ def defunct_help(cmd: str) -> str:
+    "--json",
+    is_flag=True,
+    default=False,
+    help="Prints the decoded shinylive bundle as JSON to stdout, ignoring --out.",
```
```suggestion
    help="Prints the decoded shinylive bundle as JSON to stdout, ignoring --dir.",
```

wch commented on shinylive/_main.py:
```diff
@@ -469,6 +471,153 @@ def defunct_help(cmd: str) -> str:
+    "--dir",
+    type=str,
+    default=None,
+    help="Output directory into which the app's files will be written. The directory is created if it does not exist. ",
```
Should explain what happens if `--dir` isn't used. (I think the contents get printed to stdout, right?)

wch commented on shinylive/_main.py:
```diff
@@ -469,6 +471,153 @@ def defunct_help(cmd: str) -> str:
+            if "type" in file and file["type"] == "binary":
+                print("## type: binary")
+            print("")
+            print(file["content"].encode("utf-8", errors="ignore").decode("utf-8"))
```
I just want to make sure: for binary files, this will be base64-encoded content, right? (Because it would have been base-64 encoded before being serialized to JSON and then encoded with lzstring.)

gadenbuie replied:
Yeah that's right -- this case was correct, since the `shinylive-python` or `shinylive-r` chunk includes the base64-encoded content.

wch commented on shinylive/_main.py:
```diff
@@ -469,6 +471,153 @@ def defunct_help(cmd: str) -> str:
+        os.makedirs(out_dir, exist_ok=True)
+        for file in bundle:
+            with open(out_dir / file["name"], "w") as f_out:
+                f_out.write(
+                    file["content"].encode("utf-8", errors="ignore").decode("utf-8")
+                )
```
For binary files, I think this will output base64-encoded text. I believe it needs to be base64-decoded.

(If I'm wrong about it outputting base64-encoded text, then this code would be fine but the corresponding code below, when printing to console, would be wrong -- they can't both be right.)

gadenbuie replied:
You're totally right, I missed this case. I'll update the command to write the base64-decoded binary content to the output file.

gadenbuie replied:
2da9c8ab460ae8ef3f24f6e0e0b815e2a74c1ca4

gadenbuie commented on shinylive/_url.py:
```diff
@@ -0,0 +1,217 @@
+    mode
+        The mode of the application. Defaults to "editor".
+    language
+        The language of the application. Defaults to None.
```
```suggestion
        The language of the application, or None to autodetect the language. Defaults to None.
```

gadenbuie commented on shinylive/_url.py:
```diff
@@ -0,0 +1,217 @@
+    else:
+        raise ValueError(
+            f"Language '{language}' is not supported. Please specify one of 'py' or 'r'."
+        )
```
Good point, I've kept the more general options in the CLI interface and moved the value narrowing logic into that function.

gadenbuie commented on shinylive/_url.py:
```diff
@@ -0,0 +1,217 @@
+
+def encode_shinylive_url(
+    app: str | Path,
+    files: Optional[tuple[str | Path, ...]] = None,
```
Thanks, I agree that only tuples and not lists would be awkward and I like your idea to allow passing a single file.

gadenbuie replied:
Thanks, I agree that only tuples and not lists would be awkward and I like your idea to allow passing a single file.

gadenbuie commented on shinylive/_url.py:
```diff
@@ -0,0 +1,217 @@
+class FileContentJson(TypedDict):
+    name: str
+    content: str
+    type: Literal["text", "binary"]
```
Thanks for the background and the links to the original typescript types. I've updated the imports statements and generally improved how FileContentJson objects are created to avoid adding the `type` field for text files.

gadenbuie replied:
Thanks for the background and the links to the original typescript types. I've updated the imports statements and generally improved how FileContentJson objects are created to avoid adding the `type` field for text files.

gadenbuie replied:
I added `typing_extensions` to `install_requires` but we're now running into failing checks similar to the ones that tripped us up with `lzstring`. Any ideas?

gadenbuie commented on shinylive/_url.py:
```diff
@@ -0,0 +1,217 @@
+        )
+
+    # if first file is not named either `ui.R` or `server.R`, then make it app.{language}
+    if file_bundle[0]["name"] not in ["ui.R", "server.R"]:
+        file_bundle[0]["name"] = f"app.{'py' if language == 'py' else 'R'}"
```
As it is currently written, I'm requiring that the primary app file be given to the `app` argument and then ensuring that it's the first item in the `file_bundle`.

The two lines highlighted in this comment ensure that, if the primary app file isn't `ui.R` or `server.R`, then it will be renamed `app.R` or `app.py` according to `language`. This setup is driven by shinylive requiring `app.py` or app/ui/server in R and the fact that locally we often store the primary app in a file that _isn't_ named `app.py`.

The current worst case scenario is that `ui.R` or `server.R` come first and the other file in the pair are later in the tab order, but if it really matters to someone they can make the second file the first item in `files`.

gadenbuie commented:
Thank you for your comments and suggestions @wch, they were super helpful! I've addressed and resolved them all, except for two bigger items that I'll call out because they might be hard to find in the UI above.

1. We've identified some weirdness around the dual-purpose `app` argument in `encode_shinylive_url()`. As it was reviewed, it could be a file on disk or the contents of a hypothetical `app.py` or `app.R`. I tried an alternative approach that's more consistent, but I'm worried this has user experience tradeoffs. https://github.com/posit-dev/py-shinylive/pull/20#discussion_r1443852134

2. We updated the imports to use `typing_extensions` for Python < 3.11, but this is causing issues in CI. https://github.com/posit-dev/py-shinylive/pull/20#discussion_r1443852134. In doing this, I realized the type we were adjusting already existed in this project and suffers from the same problem https://github.com/posit-dev/py-shinylive/blob/373f639ed81336f2a1393b9726d2a2a1dc48a0b9/shinylive/_app_json.py#L12-L17 https://github.com/posit-dev/py-shinylive/blob/373f639ed81336f2a1393b9726d2a2a1dc48a0b9/shinylive/_url.py#L11-L23

wch commented:
For the `typing_extensions` issue, I believe you should be able to do what we do in shiny, and list it in `install_requires`:
https://github.com/posit-dev/py-shiny/blob/592cf34397606eed98c1488799e495754ebf8a1f/setup.cfg#L34-L35

I had forgotten that that `typing_extensions` isn't part of the Python standard library, and so it needs to be listed as an explicit dependency.

gadenbuie commented:
> For the `typing_extensions` issue, I believe you should be able to do what we do in shiny, and list it in `install_requires`: [posit-dev/py-shiny@`592cf34`/setup.cfg#L34-L35](https://github.com/posit-dev/py-shiny/blob/592cf34397606eed98c1488799e495754ebf8a1f/setup.cfg#L34-L35)
> 
> I had forgotten that that `typing_extensions` isn't part of the Python standard library, and so it needs to be listed as an explicit dependency.

Did you see my [last comment above](https://github.com/posit-dev/py-shinylive/pull/20#issuecomment-1881415194)? (I know there's a lot going on in this PR its easy to miss.)

In short, it's not about `install_requires` but rather how shinylive uses a dynamic version that requires `pip` to run the module code. In other words, I found the source of the problem we paired on together last week and I have a pretty good fix for it.

wch commented:
> I've figured out the issue with `pip install -e .` and `typing_extensions`

Wait, do you mean the `lzstring` issue? (I don't recall `typing_extensions` being a problem when we paired on it last week but I could be missing something.)

gadenbuie commented:
Yeah, we hadn't added `typing_extensions` yet, but once we did we resurfaced the same problem we had with `lzstring`. Fundamentally they're caused by the same thing (see above for description). With `lzstring`, we could move the import statement into a function and delay its evaluation; with `typing_extensions` we can't use that strategy so I had toa figure out the root cause.

wch commented:
Hm, it's strange that the same thing happened with `typing_extensions` -- for shiny and a number of other packages, we use `version = attr: shiny.__version__ `, and that works fine. I think the fact that it's not working here is a symptom of something else that's weird. I'll take a deeper look into it.

gadenbuie commented:
@wch What's different about `py-shiny` is that [`__version__` is set to a literal value](https://github.com/posit-dev/py-shiny/blob/6ed505b18b3cfaab72d0a0a20762c4f0dc49b035/shiny/__init__.py#L3):

```python
__version__ = "0.6.1.9000"
```

Here in `py-shinylive` it's set a to an expression, which then forces `pip` to evaluate some of the packaged code to determine the value.

wch commented:
Ah, OK, I see. That sounds good then. Can you you rename `version` to `_version`?
